src/NetMQ/NetMQPoller.cs

```diff
@@ -112,21 +113,47 @@ protected override void QueueTask(Task task)
             m_tasksQueue.Enqueue(task);
         }
 
-        public void Run([NotNull] Action action)
+        public Task Run([NotNull] Action action)
```

Comment:
Breaking public API is not a good idea, and goes against the [C4.1 process](https://rfc.zeromq.org/spec:22/C4/) that this project uses.

I agree that returning a task would be better. It'd be better to create overloads:

```c#
public Task RunAsync([NotNull] Action action);
public Task RunAsync([NotNull] Func<CancellationToken> action, CancellationToken cancellationToken);
```

Reply:
100% agree, but was trying to avoid that because I originally thought only the ```Remove`` methods were public, and chaining only the non-public ```RunAsync``` method would avoid breaking changes...

Why is ```RunAsync``` even public?

The only reason I can see is to work around the original thread race bug when calling ```Remove()```?

Reply:
> Why is `RunAsync` even public?

`NetMQPoller` is a `TaskScheduler` and can be used to schedule work to run on the same synchronisation context (in this case: thread) as the poller. `RunAsync` is useful when you need to schedule work that needs to be on that thread, which is not an uncommon requirement.

Regardless, it's public API and must be preserved.

Reply:
> Regardless, it's public API and must be preserved.

Got that.

> RunAsync is useful when you need to schedule work that needs to be on that thread, which is not an uncommon requirement.

I think a public `RunAsync` is a design flaw.  NetMQ objects, `NetMQPoller ` in this case could and *should* handle marshalling work that *needs* to be done on the polling thread internally.  This exposes a large path for external code to break  `NetMQPoller ` and is the root source of the issue being discussed.  The only case I can think of currently *is this* case (add/remove a socket from the polled group).  Certainly, the *primary* use case for this method is the async polling of the sockets in the list for data... for which all the rest of the framework code is either internal to NetMQ or protected/private, which makes sense.